swing-lib/src/main/java/org/triplea/swing/JFrameBuilder.java

```diff
@@ -45,17 +58,32 @@ public static JFrameBuilder builder() {
    */
   public JFrame build() {
     // note: we use the two arg JFrame constructor to avoid the headless check that is in the single arg constructor.
-    final JFrame frame = new JFrame(title, null);
+    final JFrame frame = new JFrame(title, null) {
+      @Override
+      public void dispose() {
+        super.dispose();
+        Optional.ofNullable(closeAction).ifPresent(Runnable::run);
+      }
+    };
```

Comment:
I know you copied this from the original implementation, but it's still ugly. There's a Window state listener for this.

Reply:
There are enough changes here it is worth making clean IMO. I took a look and the window state listener leads to this type of code:
```
    frame.addWindowStateListener(e -> {
      if (e.getNewState() == WindowEvent.WINDOW_CLOSING) {
        Optional.ofNullable(windowCloseAction).ifPresent(Runnable::run);
      }

      if (e.getNewState() == WindowEvent.WINDOW_ACTIVATED) {
        Optional.ofNullable(windowActivatedAction).ifPresent(Runnable::run);
      }
    });
```

I do like that, except some problems do come in. After modifying things I went to add testing and found that the window state listener is only called in headed environments. The method override allows for testing to happen in headless. Weighing kinda dirty un-tested vs dirty tested code, I went with the latter. Part of the point of the swing builders is to clean up the swing API, using a "preferred" swing API to clean things up just slightly is not very consequential considering we're hiding those API's completely anyways.

For the window activated case, I'm not sure if there is a real difference between reader expectations when seeing a window adapter vs a window state listener. Given it's not a necessary update and is not clearly better, I left as-is to reduce test burden and risk.

Of substance, I added a note about why a window listener was not used.  Renaming the close action method seemed to fit well with the window activation listener, and and I added a test for the dispose listener:   f81b96453

Reply:
I get your point, but can't we just assume swing does fire its Listeners correctly?
There's a `JFrame#getWindowStateListeners` method, so we could test if `JFrameBuilder` adds the listener accordingly, and run it on our own to see if it executes the correct code at the correct events.

Reply:
> I get your point, but can't we just assume swing does fire its Listeners correctly?

Yes & no. We can't necessarily assume that the builder code adds those listeners correctly, which makes for a good suggestion to test that we do that. I do want to try and avoid the 'test-once but never again' anti-pattern that comes about when doing manual testing. For example, if we make further updates in the builder then there will be additional code that should remain 'frozen' as it is manually tested. My preference is clearly for more dirty but testable code over less dirty but not as well tested code.

Reply:
@DanVanAtta 
I think I need to clarify here, this code example (written without compiler) should make some things clear here:
```java
@Mock
private Runnable action;

@Test
void testListener() {
  final JFrame frame = JFrameBuilder.builder()
    .closeAction(action)
    .build();
  assertEquals(1, frame.getWindowStateListeners().length);
  final WindowStateListener listener = frame.getWindowStateListeners()[0];
  final WindowEvent mockedEvent = mock(WindowEvent.class);
  when(mockedEvent.getNewState()).thenReturn(WindowEvent.WINDOW_CLOSING, WindowEvent.WINDOW_ACTIVATED);
  // Test event got fired
  listener.windowStateChanged(mockedEvent);
  verify(action).run();
  // Test event doesn't get fired a second time when state type is different
  listener.windowStateChanged(mockedEvent);
  verify(action).run();
}
```
This was we only trust swing to do its job right, but test our own code completely.

Reply:
The example does help, thanks.

That test case though has a few issues:
- we go from a black-box test to a white-box test. That means higher coupling, less test efficacy, and we are more testing implementation then we are testing contract and functionality.
- we are relying on specific swing implementation details, which is in effect a white-box test into swing itself. In other words, we are hacking the swing framework and taking advantage of its poor encapsulation to accomplish this testing. Not necessarily the worst, but it's not a clean way to test.

Compared to the existing test:
```
    final AtomicBoolean testValue = new AtomicBoolean(false);

    JFrameBuilder.builder()
        .windowCloseAction(() -> testValue.set(true))
        .build()
        .dispose();

    assertThat(testValue.get(), is(true));
```
The above is *far* simpler, easier to understand, and does not depend on swing implementation as closely. We already test our code completely with the above, so I do not think it is a clear win to go to a more complicated test for the same coverage with more problems.

*But*, the technique of grabbing the window listeners is a reasonable one to test the window activation listener, which is currently not tested at all. 

https://github.com/triplea-game/triplea/pull/4653 adds a test using a similar technique for window activation.